src/main/scala/scala/Option.scala

```diff
@@ -0,0 +1,639 @@
+    * option match {
+    *   case Some(x) => f(x)
+    *   case None    => None
+    * }
+    * }}}
+    *  @param  f   the function to apply
+    *  @see map
+    *  @see foreach
+    */
+  @inline final def flatMap[B](f: A => Option[B]): Option[B] =
+    if (isEmpty) None else f(this.get)
+
+  /** Returns the nested $option value if it is nonempty.  Otherwise,
+    * return $none.
+    *
+    * This is equivalent to:
+    * {{{
+    * option match {
+    *   case Some(Some(b)) => Some(b)
+    *   case _             => None
+    * }
+    * }}}
+    * @example {{{
+    * Some(Some("something")).flatten
+    * }}}
+    *
+    * @param ev an implicit conversion that asserts that the value is
+    *           also an $option.
+    * @see flatMap
+    */
+  def flatten[B](implicit ev: A <:< Option[B]): Option[B] =
+    if (isEmpty) None else ev(this.get)
+
+  /** Returns this $option if it is nonempty '''and''' applying the predicate $p to
+    * this $option's value returns true. Otherwise, return $none.
+    *
+    * This is equivalent to:
+    * {{{
+    * option match {
+    *   case Some(x) if p(x) => Some(x)
+    *   case _               => None
+    * }
+    * }}}
+    *  @param  p   the predicate used for testing.
+    */
+  @inline final def filter(p: A => Boolean): Option[A] =
+    if (isEmpty || p(this.get)) this else None
+
+  /** Returns this $option if it is nonempty '''and''' applying the predicate $p to
+    * this $option's value returns false. Otherwise, return $none.
+    *
+    * This is equivalent to:
+    * {{{
+    * option match {
+    *   case Some(x) if !p(x) => Some(x)
+    *   case _                => None
+    * }
+    * }}}
+    *  @param  p   the predicate used for testing.
+    */
+  @inline final def filterNot(p: A => Boolean): Option[A] =
+    if (isEmpty || !p(this.get)) this else None
+
+  /** Apply the side-effecting function $f to the option's value,
+    * if it is nonempty. Otherwise, do nothing.
+    *
+    * @param f a function to apply to the option's value
+    * @tparam U the return type of f
+    * @return this Option
+    */
+  @inline final def tapEach[U](f: A => U): Option[A] = {
+    foreach(f); this
+  }
+
+  /** Returns false if the option is $none, true otherwise.
+    *
+    * This is equivalent to:
+    * {{{
+    * option match {
+    *   case Some(_) => true
+    *   case None    => false
+    * }
+    * }}}
+    *  @note   Implemented here to avoid the implicit conversion to Iterable.
+    */
+  final def nonEmpty: Boolean = isDefined
+
+  /** Necessary to keep $option from being implicitly converted to
+    *  [[scala.collection.Iterable]] in `for` comprehensions.
+    */
+  @inline final def withFilter(p: A => Boolean): WithFilter = new WithFilter(p)
+
+  /** We need a whole WithFilter class to honor the "doesn't create a new
+    *  collection" contract even though it seems unlikely to matter much in a
+    *  collection with max size 1.
+    */
+  class WithFilter(p: A => Boolean) {
```

Comment:
Is there a reason for this not to be `final`?

Reply:
I took the code of `scala.Option` from the 2.13.x branch. All your findings relate to that code. As @NthPortal points out below writing replacements for existing classes is not the way to go for this repository, so I'll provide an extension method instead (probably in a new PR, that's cleaner).